src/query/storage/m3/cluster_resolver.go

```diff
@@ -0,0 +1,299 @@
+
+	// Take any partially aggregated namespaces with longer retention or
+	// same retention with more granular resolution that may contain
+	// a matching metric.
+	for _, n := range r.partialAggregated {
+		if n.Options().Attributes().Retention > completedAttrs.Retention {
+			// Higher retention.
+			result = append(result, n)
+		} else if n.Options().Attributes().Retention == completedAttrs.Retention &&
+			n.Options().Attributes().Resolution < completedAttrs.Resolution {
+			// Same retention but more granular resolution.
+			result = append(result, n)
+		}
+	}
+
+	return namespaceCoversPartialQueryRange, result, nil
+}
+
+type reusedAggregatedNamespaceSlices struct {
+	completeAggregated []ClusterNamespace
+	partialAggregated  []ClusterNamespace
+}
+
+func (slices reusedAggregatedNamespaceSlices) reset(
+	size int,
+) reusedAggregatedNamespaceSlices {
+	// Initialize arrays if yet uninitialized.
+	if slices.completeAggregated == nil {
+		slices.completeAggregated = make([]ClusterNamespace, 0, size)
+	} else {
+		slices.completeAggregated = slices.completeAggregated[:0]
+	}
+
+	if slices.partialAggregated == nil {
+		slices.partialAggregated = make([]ClusterNamespace, 0, size)
+	} else {
+		slices.partialAggregated = slices.partialAggregated[:0]
+	}
+
+	return slices
+}
+
+// aggregatedNamespaces filters out clusters that do not meet the filter
+// condition, and organizes remaining clusters in two lists if possible.
+//
+// NB: If fanout aggregation is disabled, no clusters will be returned as either
+// partial or complete candidates. If fanout aggregation is forced to enabled
+// then no filter is applied, and all namespaces are considered viable. In this
+// case, the filter is used to determine if returned namespaces have the
+// complete set of metrics.
+//
+// NB: If fanout optimization is enabled, add any aggregated namespaces that
+// have a complete set of metrics to the completeAggregated slice list. If this
+// optimization is disabled, or if none of the aggregated namespaces are
+// guaranteed to have a complete set of all metrics, they are added to the
+// partialAggregated list.
+func aggregatedNamespaces(
+	all ClusterNamespaces,
+	slices reusedAggregatedNamespaceSlices,
+	filter func(ClusterNamespace) bool,
+	opts *storage.FanoutOptions,
+) reusedAggregatedNamespaceSlices {
+	// Reset reused slices.
+	slices = slices.reset(len(all))
+
+	// Force disable fanout to any aggregated namespaces.
+	if opts.FanoutAggregated == storage.FanoutForceDisable {
+		return slices
+	}
+
+	// Otherwise the default and force enable is to fanout and treat
+	// the aggregated namespaces differently (depending on whether they
+	// have all the data).
+	for _, namespace := range all {
+		nsOpts := namespace.Options()
+		if nsOpts.Attributes().MetricsType != storage.AggregatedMetricsType {
+			// Not an aggregated cluster.
+			continue
+		}
+
+		// If fanout is forced, ignore the filter.
+		if opts.FanoutAggregated != storage.FanoutForceEnable {
```

Comment:
I was thinking we would not actually need to skip the check here, since if `FanoutAggregatedOptimized` is forced disabled, then you'll always run this function with `filter == nil` because you won't take the path of `if len(r.completeAggregated) > 0` after running the first call to `aggregatedNamespaces(...)`.